src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,260 @@
+  // we need to check channel status.
+  // This is the time when the server will timeout for
+  // unavailability and server_unavailable_timeout_callback_
+  // will be called.
+  std::optional<absl::Time> server_unavailable_timeout_time_;
+
+  // Key is when the request will timeout and value is the request.
+  // This is only accessed in the io context thread and the destructor so
+  // no mutex is needed.
+  absl::btree_multimap<absl::Time, std::shared_ptr<RetryableGrpcRequest>>
+      pending_requests_;
+  size_t pending_requests_bytes_ = 0;
+};
+
+template <typename Service, typename Request, typename Reply>
+void RetryableGrpcClient::CallMethod(
+    PrepareAsyncFunction<Service, Request, Reply> prepare_async_function,
+    std::shared_ptr<GrpcClient<Service>> grpc_client,
+    const std::string &call_name,
+    const Request &request,
+    const ClientCallback<Reply> &callback,
+    const int64_t timeout_ms) {
+  RetryableGrpcRequest::Create(weak_from_this(),
+                               prepare_async_function,
+                               grpc_client,
+                               call_name,
+                               request,
+                               callback,
+                               timeout_ms)
+      ->CallMethod();
+}
+
+template <typename Service, typename Request, typename Reply>
+std::shared_ptr<RetryableGrpcClient::RetryableGrpcRequest>
+RetryableGrpcClient::RetryableGrpcRequest::Create(
+    std::weak_ptr<RetryableGrpcClient> weak_retryable_grpc_client,
+    PrepareAsyncFunction<Service, Request, Reply> prepare_async_function,
+    std::shared_ptr<GrpcClient<Service>> grpc_client,
+    const std::string &call_name,
+    const Request &request,
+    const ClientCallback<Reply> &callback,
+    const int64_t timeout_ms) {
+  RAY_CHECK(callback != nullptr);
+  RAY_CHECK(grpc_client.get() != nullptr);
+
+  auto executor = [weak_retryable_grpc_client,
+                   prepare_async_function,
+                   grpc_client,
+                   call_name,
+                   request,
+                   callback](std::shared_ptr<RetryableGrpcClient::RetryableGrpcRequest>
+                                 retryable_grpc_request) {
+    grpc_client->template CallMethod<Request, Reply>(
+        prepare_async_function,
+        request,
+        [weak_retryable_grpc_client, retryable_grpc_request, callback](
+            const ray::Status &status, Reply &&reply) {
+          auto retryable_grpc_client = weak_retryable_grpc_client.lock();
+          if (status.ok() || !IsGrpcRetryableStatus(status) || !retryable_grpc_client) {
+            callback(status, std::move(reply));
+            return;
+          }
+
+          retryable_grpc_client->Retry(retryable_grpc_request);
+        },
+        call_name,
+        retryable_grpc_request->GetTimeoutMs());
+  };
+
+  auto failure_callback = [callback](const ray::Status &status) {
+    callback(status, Reply());
```

Comment:
It's encouraged to use `{}` to construct variables. From effective modern C++.
One reason is, if you have a function with the same name, using `()` doesn't compile.